### PR TITLE
Sync prometheus

### PR DESCRIPTION
**Pull request by pracucci**
This PR is a follow up of https://github.com/grafana/mimir-prometheus/pull/580, synching more commits merged in Prometheus after https://github.com/grafana/mimir-prometheus/pull/580.

### CODE REVIEW

File: tsdb/compact.go [hunk 1]
```diff
@@ -170,9 +193,18 @@ func NewLeveledCompactorWithChunkSize(ctx context.Context, r prometheus.Register
 	if l == nil {
 		l = log.NewNopLogger()
 	}
+	mergeFunc := opts.MergeFunc
 	if mergeFunc == nil {
 		mergeFunc = storage.NewCompactingChunkSeriesMerger(storage.ChainedSeriesMerge)
 	}
+	var maxBlockChunkSegmentSize int64
```
Comment (pracucci) [inline tsdb/compact.go:200]: This is buggy

File: tsdb/compact.go [hunk 2]
```diff
@@ -170,9 +193,18 @@ func NewLeveledCompactorWithChunkSize(ctx context.Context, r prometheus.Register
...
+	if opts.MaxBlockChunkSegmentSize == 0 {
+		maxBlockChunkSegmentSize = chunks.DefaultChunkSegmentSize
+	}
+	var pe index.PostingsEncoder
```
Comment (pracucci) [inline tsdb/compact.go:204]: This is buggy

File: tsdb/db.go [hunk 3]
```diff
@@ -505,9 +505,7 @@ func (db *DBReadOnly) FlushWAL(dir string) (returnErr error) {
 		nil,
 		db.logger,
 		ExponentialBlockRanges(DefaultOptions().MinBlockDuration, 3, 5),
-		chunkenc.NewPool(),
-		nil,
-		false,
```
Comment (pracucci) [inline tsdb/db.go:510]: This was false

File: tsdb/compact.go [hunk 4]
```diff
@@ -181,8 +213,9 @@ func NewLeveledCompactorWithChunkSize(ctx context.Context, r prometheus.Register
 		ctx:                         ctx,
 		maxBlockChunkSegmentSize:    maxBlockChunkSegmentSize,
 		mergeFunc:                   mergeFunc,
+		postingsEncoder:             pe,
+		enableOverlappingCompaction: opts.EnableOverlappingCompaction,
```
Comment (krajorama) [inline tsdb/compact.go:217]: nice touch to reduce diff :+1:

File: tsdb/db.go (hunk 3 above)
Comment (krajorama) [reply tsdb/db.go:510]: Doesn't this (removal of setting enable overlapping blocks = false) mean that compaction works slightly differently? I cannot wrap my head around what this means on the high level. https://github.com/grafana/mimir-prometheus/blob/fc40f7cff852bf3946a6c84869539f52fe9e371e/tsdb/compact.go#L279

Comment (pracucci) [reply tsdb/db.go:510]: Yes, it's different, but this function is `DBReadOnly.FlushWAL()` and in Mimir we never use `DBReadOnly`, so it doesn't affect us.

Comment (krajorama) [reply tsdb/db.go:510]: ah, ok, now I see, in my head the option change and DBReadOnly was separate issues, approving

**Review by krajorama — APPROVED** [review]
LGTM

**Comment by pracucci** [comment]
I'm going to merge it to move forward with my work, but if anyone else has any post-merge comment please ping me directly on Slack and I will address them immediately. Thanks!

**Pull request merged**